### PR TITLE
[Improvement](Nereids) Support to query rewrite by materialized view when join input has aggregate

### PR DESCRIPTION
**Pull request by seawinde**
## Proposed changes

Support to query rewrite by materialized view when join input has aggregate, the aggregate should be simple
For example as following:
The materialized view def is 
>            select
>            l_linenumber,
>            count(distinct l_orderkey),
>           sum(case when l_orderkey in (1,2,3) then l_suppkey * l_linenumber else 0 end),
>            max(case when l_orderkey in (4, 5) then (l_quantity *2 + part_supp_a.qty_max) * 0.88 else 100 end),
>            avg(case when l_partkey in (2, 3, 4) then l_discount + o_totalprice + part_supp_a.qty_sum else 50 end)
>            from lineitem
>            left join orders on l_orderkey = o_orderkey
>            left join 
 >           (select ps_partkey, ps_suppkey, sum(ps_availqty) qty_sum, max(ps_availqty) qty_max,
 >               min(ps_availqty) qty_min,
 >               avg(ps_supplycost) cost_avg
 >               from partsupp
 >               group by ps_partkey,ps_suppkey) part_supp_a
 >           on l_partkey = part_supp_a.ps_partkey
 >           and l_suppkey = part_supp_a.ps_suppkey
 >           group by l_linenumber;

when query is like following, it can be rewritten by mv above
>                        select
>            l_linenumber,
>            sum(case when l_orderkey in (1,2,3) then l_suppkey * l_linenumber else 0 end),
>            avg(case when l_partkey in (2, 3, 4) then l_discount + o_totalprice + part_supp_a.qty_sum else 50 end)
>            from lineitem
>            left join orders on l_orderkey = o_orderkey
>            left join 
>            (select ps_partkey, ps_suppkey, sum(ps_availqty) qty_sum, max(ps_availqty) qty_max,
>                min(ps_availqty) qty_min,
>                avg(ps_supplycost) cost_avg
>                from partsupp
>                group by ps_partkey,ps_suppkey) part_supp_a
>            on l_partkey = part_supp_a.ps_partkey
>            and l_suppkey = part_supp_a.ps_suppkey
>            group by l_linenumber;

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/joinorder/hypergraph/node/StructInfoNode.java [hunk 1]
```diff
@@ -65,24 +67,49 @@ public StructInfoNode(int index, List<HyperGraph> graphs) {
...
+        plan.accept(new DefaultPlanVisitor<Void, Pair<Boolean, ImmutableList.Builder<Set<Expression>>>>() {
+            @Override
+            public Void visitLogicalAggregate(LogicalAggregate<? extends Plan> aggregate,
+                    Pair<Boolean, ImmutableList.Builder<Set<Expression>>> collector) {
+                if (!collector.key()) {
+                    return null;
+                }
+                collector.value().add(ImmutableSet.copyOf(aggregate.getExpressions()));
+                collector.value().add(ImmutableSet.copyOf(((LogicalAggregate<?>) plan).getGroupByExpressions()));
+                return super.visit(aggregate, collector);
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/joinorder/hypergraph/node/StructInfoNode.java:81]: visit child first

**Review by keanji-x — APPROVED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/joinorder/hypergraph/node/StructInfoNode.java [hunk 2]
```diff
@@ -65,24 +67,49 @@ public StructInfoNode(int index, List<HyperGraph> graphs) {
...
+                return super.visit(filter, collector);
+            }
+
+            @Override
+            public Void visitGroupPlan(GroupPlan groupPlan,
+                    Pair<Boolean, ImmutableList.Builder<Set<Expression>>> collector) {
+                if (!collector.key()) {
+                    return null;
+                }
+                return super.visit(groupPlan.getGroup().getLogicalExpression().getPlan(), collector);
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/joinorder/hypergraph/node/StructInfoNode.java:100]:
```suggestion
                groupPlan.getGroup().getLogicalExpression().get(0).getPlan().accept(this, collector);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/joinorder/hypergraph/node/StructInfoNode.java [hunk 3]
```diff
@@ -65,24 +67,49 @@ public StructInfoNode(int index, List<HyperGraph> graphs) {
     }
 
     private @Nullable List<Set<Expression>> collectExpressions(Plan plan) {
-        if (plan instanceof LeafPlan) {
-            return ImmutableList.of();
-        }
-        List<Set<Expression>> childExpressions = collectExpressions(plan.child(0));
-        if (!isValidNodePlan(plan) || childExpressions == null) {
-            return null;
-        }
```
Comment (XieJiann) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/joinorder/hypergraph/node/StructInfoNode.java:74]: if (plan instance GroupPlan) {
  res = collect(plan.getLogcialExpr().get(0).get(plan));
}

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Pull request merged**